streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala

```diff
@@ -365,28 +365,28 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    *     .leftJoin(userRegionsTable, (clicks: Long, region: String) => (if (region == null) "UNKNOWN" else region, clicks))
    *     .map((_, regionWithClicks) => regionWithClicks)
    *
-   *     // the groupByKey gets the Serialized instance through an implicit conversion of the
+   *     // the groupByKey gets the Grouped instance through an implicit conversion of the
    *     // serdes brought into scope through the import Serdes._ above
    *     .groupByKey
    *     .reduce(_ + _)
    *
-   * // Similarly you can create an implicit Serialized and it will be passed implicitly
+   * // Similarly you can create an implicit Grouped and it will be passed implicitly
    * // to the groupByKey call
    * }}}
    *
-   * @param serialized the instance of Serialized that gives the serdes
+   * @param grouped the instance of Grouped that gives the serdes
    * @return a [[KGroupedStream]] that contains the grouped records of the original [[KStream]]
    * @see `org.apache.kafka.streams.kstream.KStream#groupByKey`
    */
-  def groupByKey(implicit serialized: Serialized[K, V]): KGroupedStream[K, V] =
-    inner.groupByKey(serialized)
+  def groupByKey(implicit grouped: Grouped[K, V]): KGroupedStream[K, V] =
```

Comment:
This is my only semi-major concern.

Reply:
tl;dr: we can't do this in a way that breaks no one. We have to choose between two populations of users to break.

After actually trying this, we face the following dilemma:

1. **if our users are importing the implicit conversions with a wildcard** (`import org.apache.kafka.streams.scala.ImplicitConversions._`), **then having both overloads in `KStream` will break them.**
This is because the compiler says "I have some serdes. I need either a `Grouped` or a `Serialized`. I know how to make both `Grouped` and `Serialized` from the serdes I have. Therefore, I need the programmer to resolve this ambiguity."
**On the other hand, if we just drop the Serialized overload, their code will continue to work**, since the compiler will just use the implicit conversion to produce a Grouped instead of a Serialized.

2. **if they are only importing the implicit conversions they need** (`import org.apache.kafka.streams.scala.ImplicitConversions.serializedFromSerde`), **then dropping the Serialized overload will break them**.
This is because the compiler only knows how to implicitly produce a Serialized, not a Grouped.
On the other hand, **if we keep both overloads, their code will continue to work**, and in fact, they'll get a deprecation warning to switch to grouped, which they could do by just omitting `serializedFromSerde` and importing `groupedFromSerde` instead.

Just to preach for a second, this is why it's dangerous to use wildcard imports with implicit conversions... you have no idea what is actually happening. Implicits are nice for saving boilerplate, not for saving you from knowing what your program is doing.

So my opinionated preference is that our users should be going with the specific imports (2), and therefore we should do the same thing with the scala API as we do with the java one: deprecate old methods.

But pragmatically, I think the fact that the original contributors used the wildcard pattern (1) suggests that our scala users in general are more likely to also be using pattern 1. In turn, this suggests we should just go ahead and replace the Serialized method with the Grouped one.

Reply:
I've been thinking about this and just yanking out `Serialized` with no prior warning doesn't feel like it's the correct thing to do, so I have an another proposal

1. I'll submit another minor PR where we deprecate `Serialized` in the Scala API.  This PR should get merged to trunk and cherry-picked to 2.1 and go with the 2.1 release. Also, update the docs stating we'll remove `Serialized` (from the Scala API) in the next release.

2. Update this PR to remove `Serialized` (per @vvcephei comments above) and associated methods and tests. Then merge it after the code freeze, so that it ships with 2.2.  Then we will have provided one release "warning" so to speak.

IMHO it's import to take these steps now, as it's too easy to let the Scala API "drift" some.

WDYT?

\cc @guozhangwang @mjsax @vvcephei

Reply:
I agree. Scala API is public API same as Java. Deprecating `Serialized` in Scala should be part of the KIP (Bill, can you update it?) -- I think, we need to be more careful in the future and include Scala API impact in all KIP discussions (guess, Scala API is still too new, so we tend to let it slip, what we should do better in the future).

I am just not sure, if we can/should remove the `Serialized` in 2.2 already. In the past, we used only mayor releases to remove deprecate APIs, so for this, it might be 3.0 (I say 'might' because, we should keep deprecates API for a year, too, so only remove if both conditions are met).

Thoughts?

Reply:
@mjsax - I agree with your thoughts on keeping the deprecated API until  3.0. And yes I can update the KIP to mention Scala API deprecation.

But as I look at the bigger picture, the driving factor behind adding `Grouped` is that users can name repartition topics and then not break the ability to do a rolling upgrade when adding or removing operations from their topology.

IMHO by not putting in `Grouped` we are unnecessarily penalizing Scala API users (and I realize it was my suggestion just to deprecate `Serialized` and leave out `Grouped` in this release).

So I'm going to amend my proposal to this:

1. Deprecate `Serialized` as planned
2. Add the `Grouped` updates to the Scala API.
3. Per @vvcephei original suggestions we explicitly list imports concerning implicit converters which will resolve the ambiguous method reference errors.  And to preserve compatibility, we'd include `serializedFromSerde` and not import `groupedFromSerde`.

I _think_ this approach will work because if a user wants to name a repartition topic, she will have to provide a `Grouped.as("name")` argument (the name can 't be provided implicitly) in addition to importing `groupedFromSerde`.

At that point importing both implicit converters should not cause any ambiguity as `Serialized` is chosen for no args and `Grouped` is used due to the multiple parameter lists of the function  `Grouped.with[K, V](name: String)(implicit keySerde: Serde[K], valueSerde: Serde[V])`.

I'll verify this approach with some updated tests; then we can include everything in this PR for 2.1 and remove `Serialized` in a future release when deeming necessary.  Also, I can update the documentation to explain this situation to users and not to

Thoughts?

Reply:
> Add the Grouped updates to the Scala API.

Absolutely.

(Don't now enough about Scala to comment on the rest. Leave this part to @guozhangwang and @vvcephei.)

Reply:
Update I've got something working that will work for everyone I believe, and I'll post a PR tomorrow